### ISSUE TITLE
Does not work in JupyterLab

### ISSUE THREAD

**Issue by smackesey** [glacambre/firenvim#622]
- OS Version: mac x86-64
- Browser Version: Google Inc. Chrome 83.0.4103.116
- Browser Addon Version: 0.1.29
- Neovim Plugin Version: 0.1.29

- What I tried to do: use FireNvim in JupyterLab
- What happened: An nvim instance briefly starts up but then quits immediately.

Firenvim works for me in traditional [Jupyter notebooks](), but not in the JupyterLab app. When I click on a cell in a notebook in JupyterLab, the cell flashes briefly but then quits. Here is the error backtrace from console:

    12:51:54.491 chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6521 Error: Cannot read property 'getValue' of undefined
        at wrappedSendMessageCallback (chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6156)
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6521
    async function (async)
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6308
    load (async)
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6302
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6301
    ./src/NeovimFrame.ts @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6528
    __webpack_require__ @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:20
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:84
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:87
    12:51:54.509 chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6523 Uncaught (in promise) undefined
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6523
    async function (async)
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6308
    load (async)
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6302
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6301
    ./src/NeovimFrame.ts @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:6528
    __webpack_require__ @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:20
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:84
    (anonymous) @ chrome-extension://egpjdkipkomnmjhjmdamaniclmdlobbo/nvimui.js:87

I also know that NeoVim is starting up, as my init file is executed (tested with some echos). So it appears something is triggering NeoVim to quit immediately. Not sure if its relevant, but JupyterLab uses CodeMirror as a cell editor.

**Comment by glacambre** [comment]
Thanks for the detailed bug report! I can reproduce this bug - I'll investigate :).

**Comment by glacambre** [comment]
This happens because the firenvim frame is attached to the element you click on - but the page removes that element from the document right after you click on it. Since the element the firenvim frame is attached to is removed, Firenvim removes itself from the page.

The problem is that even if this feature was removed, we'd still have the issue of syncing Neovim's buffer with the page's element. Since the element the Firenvim frame was attached to doesn't exist anymore, Firenvim won't know where to put its data.

I'm not sure how to fix this. There's probably no other way than writing a jupyterlab-specific editor handler. Doing this will probably take me a while - I'm currently focusing on #347 :).

**Comment by smackesey** [comment]
Thanks @glacambre for your prompt response. This issue is important to me-- vim embedded in a full-featured data science environment would be extremely useful for many. I may be willing to take a crack at the jupyterlab-specific handler myself if you can give me a few pointers. Are there other site-specific handlers in place that I could take as a starting point?

**Comment by glacambre** [comment]
I'd be very happy to merge your work :)

There are two sides to Firenvim - one is the Neovim plugin and the other is the browser plugin. You'll only need to implement changes to the browser plugin. In order to do that, install nodejs and npm, clone the firenvim github repository somewhere, go to the cloned directory and run `npm install`. This should install all needed dependencies.

Then run `"$(npm bin)/webpack" chrome` (or `npm run build` if you want to build for both firefox and chrome) - this should create a directory named `target/chrome`, which contains a file named `manifest.json`. Then, in your browser, go to `chrome://extensions`, remove your current installation of Firenvim and click on `load unpacked`. Navigate to and select the directory named `target/chrome`. This should load Firenvim in your browser - go to `https://txti.es` and make sure it works.

Every time you'll make changes to Firenvim you'll need to reload it - to do this, go to `chrome://extensions` and click on the refresh button (the arrow that makes a circle) and then go back to the tab you were in and reload it.

Once you're done setting things up, you're ready to hack:

You'll need to implement a new [editor](https://github.com/glacambre/firenvim/blob/master/src/editors/AbstractEditor.ts). There already are a few editors ([textareas](https://github.com/glacambre/firenvim/blob/master/src/editors/TextareaEditor.ts), [codemirror](https://github.com/glacambre/firenvim/blob/master/src/editors/CodeMirrorEditor.ts)…) you can take inspiration from.

Basically, this new editor would need to implement a [matches](https://github.com/glacambre/firenvim/blob/45188db781bb7d158d09d061f71516c85117e68b/src/editors/CodeMirrorEditor.ts#L7-L18) function that will detect whether an element is a jupyterlab input element - ideally by using information present on the element, information present in the page if not possible and by checking document.location.href if there really is no other solution.

After that, you'll need to add your JupyterLabEditor to the list of editors that Firenvim uses [here](https://github.com/glacambre/firenvim/blob/45188db781bb7d158d09d061f71516c85117e68b/src/editors/editors.ts#L9).

Then, you'll need to implement the constructor of that editor. This is probably the trickiest part of the endeavor - you'll have to find a way to compute a reference to the element the content of which you are going to get/set.

Once this is done, you can start implementing the rest of the [AbstractEditor](https://github.com/glacambre/firenvim/blob/master/src/editors/AbstractEditor.ts). The only functions that are absolutely needed are getElement, setContent and getContent - it's ok to return default values for the rest (for example the Monaco editor doesn't have a way to [get](https://github.com/glacambre/firenvim/blob/45188db781bb7d158d09d061f71516c85117e68b/src/editors/MonacoEditor.ts#L44-L46) or [set](https://github.com/glacambre/firenvim/blob/45188db781bb7d158d09d061f71516c85117e68b/src/editors/MonacoEditor.ts#L72-L74) the cursor and the Textarea editor doesn't provide a way to [get the buffer's language](https://github.com/glacambre/firenvim/blob/45188db781bb7d158d09d061f71516c85117e68b/src/editors/TextareaEditor.ts#L47-L49)).

Once you've done all of that, things should work - Firenvim will take care of the rest. If you need to debug things, take a look at chrome's debugger (`<C-S-i>` by default).

Feel free to ask questions, I'll be happy to answer them :)

**Comment by smackesey** [comment]
Wow, thank you for the detailed intstructions. Just got the dev version running and starting to hack now. A few comments to help anyone else who wants to do a PR:

- the `$(npm bin)/webpack chrome` command produced an error (pasted below), but `npm run build` worked
- after performing the above steps, I had to restart Chrome to get it working. Before restart, txti.es would throw an error.

**Comment by smackesey** [comment]
OK, I think I fixed the problem with 6 characters of code. The CodeMirrorEditor was being triggered in JupyterLab, but the element match was not working correctly because JupyterLab contains elements with class `jp-CodeMirrorEditor`, which was being matched by the `CodeMirrorEditor` element detection routine. I modified the regexp used in this routine to correctly detect the underlying CodeMirror editor element. See my PR.
